File: polyglot/csharp/network_policy_analyzer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;

namespace bastionkit.polyglot.csharp
{
    // ============================================================================
    // DOMAIN MODELS
    // ============================================================================

    public interface IPolicyDocument
    {
        string? Namespace { get; }
        List<PolicyRule>? IngressRules { get; }
        List<PolicyRule>? EgressRules { get; }
        PolicyPriority Priority { get; }
    }

    public class PolicyRule
    {
        public List<string>? FromPorts { get; set; }
        public List<string>? ToPorts { get; set; }
        public string? Protocol { get; set; } // TCP, UDP, *
        public bool Allow { get; set; } = true;
        public List<string>? Sources { get; set; }
        public List<string>? Destinations { get; set; }
    }

    public enum PolicyPriority : byte
    {
        Critical,
        High,
        Medium,
        Low,
        Info
    }

    // ============================================================================
    // SECURITY BASELINE CONFIGURATION
    // ============================================================================

    public class SecurityBaseline
    {
        public const int DefaultDenyThreshold = 1;
        public const int MaxIngressPortsPerRule = 5;
        public const int MaxEgressRulesPerNamespace = 20;
        public static readonly string[] AllowedProtocols = { "TCP", "UDP", "*", "SCTP" };

        private SecurityBaseline() { }

        public static SecurityBaseline Instance => new();
   

        // ============================================================================
        // CORE ANALYZER INTERFACE
        // ============================================================================

        public interface INetworkPolicyAnalyzer
        {
            AnalysisResult Analyze(IPolicyDocument document);
            void SetBaseline(SecurityBaseline? baseline = null);
        }

        // ============================================================================
        // IMPLEMENTATION: NETWORK POLICY ANALYZER
        // ============================================================================

        public class NetworkPolicyAnalyzer : INetworkPolicyAnalyzer
        {
            private readonly SecurityBaseline _baseline;
            private readonly List<AnalysisFinding> _findings;

            public NetworkPolicyAnalyzer(SecurityBaseline? baseline = null)
            {
                _baseline = baseline ?? SecurityBaseline.Instance;
                _findings = new();
            }

            public AnalysisResult Analyze(IPolicyDocument document)
            {
                var result = new AnalysisResult(document.Namespace);
                result.Document = document;
                result.Score = 100.0;

                // Phase 1: Structural checks
                CheckDefaultDeny(result, document);
                CheckRuleComplexity(result, document);

                // Phase 2: Ingress analysis
                AnalyzeIngressRules(result, document.IngressRules);

                // Phase 3: Egress analysis  
                AnalyzeEgressRules(result, document.EgressRules);

                // Phase 4: Cross-checks and recommendations
                CheckCrossNamespaceTraffic(result, document);

                // Calculate final score
                result.Score = CalculateScore(result);

                return result;
            }

            public void SetBaseline(SecurityBaseline? baseline)
            {
                _baseline = baseline ?? SecurityBaseline.Instance;
            }

        // ============================================================================
        // ANALYSIS LOGIC - PHASE 1: STRUCTURAL CHECKS
        // ============================================================================

        private static void CheckDefaultDeny(AnalysisResult result, IPolicyDocument doc)
        {
            var hasIngressDeny = doc.IngressRules?.Any(r => !r.Allow) ?? false;
            var hasEgressDeny = doc.EgressRules?.Any(r => !r.Allow) ?? false;

            if (!hasIngressDeny && !hasEgressDeny)
            {
                result.Findings.Add(new AnalysisFinding(
                    PolicyPriority.Critical,
                    "Missing default deny policy - namespace allows all traffic by default",
                    new[] { "Add 'allow: false' rule at end of both Ingress and Egress sections" },
                    10.0));
            }

            if (hasIngressDeny && hasEgressDeny)
            {
                result.Findings.Add(new AnalysisFinding(
                    PolicyPriority.High,
                    "Default deny policies present - good baseline",
                    new[] { "Verify rules are ordered correctly (deny last)" },
                    5.0));
            }
        }

        private static void CheckRuleComplexity(AnalysisResult result, IPolicyDocument doc)
        {
            if (doc.IngressRules?.Count > _baseline.MaxEgressRulesPerNamespace)
            {
                result.Findings.Add(new AnalysisFinding(
                    PolicyPriority.Medium,
                    $"Ingress rule count ({doc.IngressRules.Count}) exceeds threshold ({_baseline.MaxEgressRulesPerNamespace})",
                    new[] { "Consolidate rules using CIDR ranges or port groups" },
                    3.0));
            }

            if (doc.EgressRules?.Count > _baseline.MaxEgressRulesPerNamespace)
            {
                result.Findings.Add(new AnalysisFinding(
                    PolicyPriority.Medium,
                    $"Egress rule count ({doc.EgressRules.Count}) exceeds threshold",
                    new[] { "Group egress by service tier or application" },
                    3.0));
            }
        }

        // ============================================================================
        // ANALYSIS LOGIC - PHASE 2: INGRESS ANALYSIS
        // ============================================================================

        private static void AnalyzeIngressRules(AnalysisResult result, List<PolicyRule>? rules)
        {
            if (rules == null || !rules.Any()) return;

            var overlyPermissive = new List<(int Index, PolicyRule Rule)>();

            foreach (var (index, rule) in rules.Select((r, i) => (i, r)))
            {
                // Check for overly broad source ranges
                if (!string.IsNullOrEmpty(rule.FromPorts?.FirstOrDefault()))
                {
                    var portCount = rule.FromPorts.Count;
                    if (portCount > _baseline.MaxIngressPortsPerRule)
                    {
                        result.Findings.Add(new AnalysisFinding(
                            PolicyPriority.Low,
                            $"Ingress rule #{index + 1} has {portCount} ports - consider splitting",
                            new[] { "Split into multiple rules with specific port ranges" },
                            1.0));
                    }

                    // Check for wildcard protocols
                    if (rule.Protocol != null && !SecurityBaseline.AllowedProtocols.Contains(rule.Protocol))
                    {
                        result.Findings.Add(new AnalysisFinding(
                            PolicyPriority.Medium,
                            $"Ingress rule #{index + 1} uses non-standard protocol: {rule.Protocol}",
                            new[] { "Restrict to TCP/UDP/* or explicitly allowed protocols" },
                            4.0));
                    }

                    // Check for overly permissive sources (e.g., 0.0.0.0/0)
                    if (!string.IsNullOrEmpty(rule.Sources?.FirstOrDefault()))
                    {
                        var source = rule.Sources.First();
                        if (source.Contains("0.0.0.0") || source == "*")
                        {
                            result.Findings.Add(new AnalysisFinding(
                                PolicyPriority.High,
                                $"Ingress rule #{index + 1} allows traffic from any source: {source}",
                                new[] { "Restrict to specific CIDR blocks or namespaces" },
                                8.0));
                        }
                    }
                }

                // Check for missing port specification (allows all ports)
                if (rule.FromPorts == null || rule.FromPorts.Count == 0)
                {
                    result.Findings.Add(new AnalysisFinding(
                        PolicyPriority.Medium,
                        $"Ingress rule #{index + 1} has no port restrictions",
                        new[] { "Add 'fromPorts' field with specific ports or ranges" },
                        5.0));
                }
            }

            // Check if any ingress allows all traffic (no source/port limits)
            var unrestrictedIngress = rules.Any(r => 
                r.FromPorts == null || !r.FromPorts.Any());

            if (unrestrictedIngress)
            {
                result.Findings.Add(new AnalysisFinding(
                    PolicyPriority.Critical,
                    "At least one ingress rule allows all traffic",
                    new[] { "Add port restrictions or source limits to every ingress rule" },
                    15.0));
            }
        }

        // ============================================================================
        // ANALYSIS LOGIC - PHASE 3: EGRESS ANALYSIS
        // ============================================================================

        private static void AnalyzeEgressRules(AnalysisResult result, List<PolicyRule>? rules)
        {
            if (rules == null || !rules.Any()) return;

            var unrestrictedEgress = new List<(int Index, PolicyRule Rule)>();

            foreach (var (index, rule) in rules.Select((r, i) => (i, r)))
            {
                // Check for unrestricted egress
                if (rule.ToPorts == null || !rule.ToPorts.Any())
                {
                    result.Findings.Add(new AnalysisFinding(
                        PolicyPriority.Critical,
                        $"Egress rule #{index + 1} allows all traffic to any destination",
                        new[] { "Add 'toPorts' and 'destinations' restrictions" },
                        20.0));

                    unrestrictedEgress.Add((index, rule));
                }

                // Check for overly broad destinations
                if (!string.IsNullOrEmpty(rule.Destinations?.FirstOrDefault()))
                {
                    var dest = rule.Destinations.First();
                    if (dest.Contains("0.0.0.0") || dest == "*")
                    {
                        result.Findings.Add(new AnalysisFinding(
                            PolicyPriority.High,
                            $"Egress rule #{index + 1} allows traffic to any destination: {dest}",
                            new[] { "Restrict to required external services only" },
                            10.0));
                    }
                }

                // Check for missing egress restrictions
                if (rule.ToPorts == null || !rule.ToPorts.Any())
                {
                    result.Findings.Add(new AnalysisFinding(
                        PolicyPriority.Medium,
                        $"Egress rule #{index + 1} has no port restrictions",
                        new[] { "Add 'toPorts' field with specific ports" },
                        4.0));
                }

                // Check for external egress (potential data exfiltration)
                if (!string.IsNullOrEmpty(rule.Protocol))
                {
                    var isExternal = rule.Destinations?.Any(d => 
                        d.Contains(".") && !d.StartsWith("127.") && !d.StartsWith("10.") && 
                        !d.StartsWith("192.168.") && !d.StartsWith("172."));

                    if (isExternal)
                    {
                        result.Findings.Add(new AnalysisFinding(
                            PolicyPriority.High,
                            $"Egress rule #{index + 1} allows external traffic",
                            new[] { "Document business justification for each external dependency" },
                            6.0));
                    }
                }
            }

            if (unrestrictedEgress.Any())
            {
                result.Findings.Add(new AnalysisFinding(
                    PolicyPriority.Critical,
                    $"{unrestrictedEgress.Count} egress rules allow unrestricted traffic",
                    new[] { "Apply least-privilege principle to external dependencies" },
                    25.0));
            }
        }

        // ============================================================================
        // ANALYSIS LOGIC - PHASE 4: CROSS-CHECKS
        // ============================================================================

        private static void CheckCrossNamespaceTraffic(AnalysisResult result, IPolicyDocument doc)
        {
            if (doc.Namespace == null || string.IsNullOrEmpty(doc.Namespace))
            {
                result.Findings.Add(new AnalysisFinding(
                    PolicyPriority.Medium,
                    "Missing namespace identifier in policy document",
                    new[] { "Add 'namespace' field to properly scope the policy" },
                    2.0));
            }

            // Check for potential cross-namespace traffic patterns
            if (doc.IngressRules?.Any(r => r.Allow) == true && 
                doc.EgressRules?.Any(r => r.Allow) == true)
            {
                result.Findings.Add(new AnalysisFinding(
                    PolicyPriority.Info,
                    "Policy allows both ingress and egress traffic",
                    new[] { "Verify this is intentional for the workload type" },
                    0.5));
            }

            // Check for potential data exfiltration paths
            var externalEgressCount = doc.EgressRules?.Count(r => 
                r.ToPorts != null && r.ToPorts.Any()) ?? 0;

            if (externalEgressCount > 3)
            {
                result.Findings.Add(new AnalysisFinding(
                    PolicyPriority.Medium,
                    $"High egress complexity: {externalEgressCount} external egress rules",
                    new[] { "Review each rule for necessity and consolidation opportunities" },
                    2.0));
            }
        }

        // ============================================================================
        // SCORING AND REPORTING
        // ============================================================================

        private static double CalculateScore(AnalysisResult result)
        {
            if (result.Findings.Count == 0) return 100.0;

            var totalPenalty = result.Findings.Sum(f => f.Penalty);
            var score = Math.Max(0, 100.0 - totalPenalty);

            // Apply severity multiplier for critical findings
            var criticalCount = result.Findings.Count(f => f.Priority == PolicyPriority.Critical);
            if (criticalCount > 0)
            {
                score -= criticalCount * 5;
            }

            return Math.Max(0, score);
        }

        // ============================================================================
        // PUBLIC API: ANALYSIS RESULT
        // ============================================================================

        public class AnalysisResult
        {
            public string? Namespace { get; set; }
            public IPolicyDocument? Document { get; set; }
            public double Score { get; set; }
            public List<AnalysisFinding> Findings { get; } = new();
            public DateTime AnalyzedAt { get; set; } = DateTime.UtcNow;

            public AnalysisResult(string? namespaceName)
            {
                Namespace = namespaceName;
            }

            public void AddFinding(AnalysisFinding finding)
            {
                Findings.Add(finding);
            }

            public bool HasCriticalIssues() => 
                Findings.Any(f => f.Priority == PolicyPriority.Critical);

            public bool IsCompliant(double threshold = 80.0) => Score >= threshold;

            public string GenerateSummary()
            {
                var criticalCount = Findings.Count(f => f.Priority == PolicyPriority.Critical);
                var highCount = Findings.Count(f => f.Priority == PolicyPriority.High);
                var mediumCount = Findings.Count(f => f.Priority == PolicyPriority.Medium);

                return $"""
                    Namespace: {Namespace ?? "Unknown"}
                    Security Score: {Score:F1}%
                    
                    Severity Summary:
                      - Critical: {criticalCount}
                      - High:    {highCount}
                      - Medium:  {mediumCount}
                      - Low/Info:{Findings.Count(f => f.Priority != PolicyPriority.Critical && 
                                              f.Priority != PolicyPriority.High && 
                                              f.Priority != PolicyPriority.Medium)}

                    Compliance Status: {(IsCompliant() ? "COMPLIANT" : "NON-COMPLIANT")}
                    
                    Top Issues:
                """;
            }
        }

        public class AnalysisFinding
        {
            public PolicyPriority Priority { get; set; }
            public string Title { get; set; } = "";
            public List<string> Recommendations { get; set; } = new();
            public double Penalty { get; set; }

            public AnalysisFinding(PolicyPriority priority, string title, 
                                   List<string>? recommendations = null, 
                                   double penalty = 0.0)
            {
                Priority = priority;
                Title = title;
                Recommendations ??= new();
                if (recommendations != null)
                {
                    foreach (var rec in recommendations)
                    {
                        Recommendations.Add(rec);
                    }
                }
                Penalty = penalty;
            }

            public string ToMarkdown() => $"""
                    ### [{Priority}] {Title}
                    
                    **Penalty:** -{Penalty:F1}%
                    
                    

File: polyglot/python/network_policy_analyzer.py
"""
polyglot/python/network_policy_analyzer.py

Hardened security baseline for air-gapped/regulated Kubernetes.
Assess and generate NetworkPolicy/PodSecurity/RBAC/quota controls.

This module provides:
- YAML parsing of existing policies
- Validation against best practices (deny-all default, least privilege)
- Gap analysis - what's missing vs recommended
- Generation of remediation policies
"""

import json
import re
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum, auto
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple, Set


# =============================================================================
# CONSTANTS & CONFIGURATION
# =============================================================================

DEFAULT_DENY_ALL_PORTS = [80, 443, 53]  # HTTP/HTTPS/DNS - default allow ports
DEFAULT_DENY_ALL_NAMESPACES = ["kube-system", "kube-public"]
MIN_RECOMMENDED_POLICY_COUNT = 1
MAX_RULES_PER_NAMESPACE = 20

# PodSecurity Standards (Kubernetes 1.24+)
POD_SECURITY_STANDARDS = {
    "restricted": {"runAsNonRoot": True, "allowPrivilegeEscalation": False},
    "baseline": {"runAsNonRoot": False, "allowPrivilegeEscalation": True},
}

# RBAC Best Practices
RBAC_MIN_GROUPS_REQUIRED = ["view", "edit"]


# =============================================================================
# DATA MODELS
# =============================================================================

class PolicyStatus(Enum):
    """Status of a policy rule after analysis."""
    COMPLIANT = auto()
    WARNING = auto()
    ERROR = auto()
    INFO = auto()


@dataclass(frozen=True)
class PortRule:
    """Represents a port-based network rule."""
    name: str
    protocol: str  # "TCP", "UDP", or "*"
    port: int
    action: str  # "Allow" or "Deny"
    from_namespaces: List[str] = field(default_factory=list)
    to_namespaces: List[str] = field(default_factory=list)
    
    @property
    def is_default_port(self) -> bool:
        return self.port in DEFAULT_DENY_ALL_PORTS


@dataclass(frozen=True)
class NamespaceRule:
    """Represents a namespace-based network rule."""
    name: str
    action: str  # "Allow" or "Deny"
    from_namespaces: List[str] = field(default_factory=list)
    to_namespaces: List[str] = field(default_factory=list)


@dataclass(frozen=True)
class PolicyRule:
    """Represents a single rule within a NetworkPolicy."""
    name: str
    kind: str  # "Ingress" or "Egress"
    ports: List[PortRule] = field(default_factory=list)
    from_namespaces: List[NamespaceRule] = field(default_factory=list)
    to_namespaces: List[NamespaceRule] = field(default_factory=list)


@dataclass(frozen=True)
class NetworkPolicy:
    """Represents a Kubernetes NetworkPolicy object."""
    name: str
    namespace: str
    kind: str  # "NetworkPolicy" or "PodSecurityPolicy"
    spec: Dict[str, Any] = field(default_factory=dict)
    rules: List[PolicyRule] = field(default_factory=list)
    
    @property
    def has_default_deny(self) -> bool:
        """Check if policy has a default deny rule."""
        ingress_rules = [r for r in self.rules if r.kind == "Ingress"]
        egress_rules = [r for r in self.rules if r.kind == "Egress"]
        
        # Check for explicit deny-all rules
        for rule in ingress_rules + egress_rules:
            if not rule.ports and not rule.from_namespaces:
                return True
        
        return False


@dataclass(frozen=True)
class AnalysisResult:
    """Container for analysis results."""
    policies_analyzed: int = 0
    total_rules: int = 0
    compliant_policies: int = 0
    warning_count: int = 0
    error_count: int = 0
    
    findings: List[Dict[str, Any]] = field(default_factory=list)
    recommendations: List[Dict[str, Any]] = field(default_factory=list)
    
    @property
    def score(self) -> float:
        """Calculate compliance score (0-100)."""
        total = self.compliant_policies + self.warning_count + max(0, self.error_count - 5)
        if total == 0:
            return 100.0
        
        compliant_ratio = self.compliant_policies / total
        warning_penalty = min(self.warning_count * 2, 30)
        error_penalty = min(self.error_count * 5, 40)
        
        score = max(0, (compliant_ratio * 100) - warning_penalty - error_penalty)
        return round(score, 1)


# =============================================================================
# YAML PARSING & LOADING
# =============================================================================

def load_policy_from_yaml(yaml_content: str) -> Optional[NetworkPolicy]:
    """Parse a single NetworkPolicy from YAML content."""
    try:
        import yaml
        
        data = yaml.safe_load(yaml_content)
        
        if not isinstance(data, dict):
            return None
            
        kind = data.get("kind", "")
        if kind != "NetworkPolicy":
            return None
            
        name = data.get("metadata", {}).get("name", "unnamed")
        namespace = data.get("metadata", {}).get("namespace", "default")
        
        spec = data.get("spec", {})
        policy_rules = []
        
        # Parse ingress rules
        for rule in spec.get("ingress", []):
            ports = parse_port_list(rule.get("ports", []))
            from_ns = parse_namespace_list(rule.get("from", []))
            
            policy_rules.append(PolicyRule(
                name=f"{name}-ingress-{rule.get('name', 'unnamed')}",
                kind="Ingress",
                ports=ports,
                from_namespaces=from_ns,
                to_namespaces=[]
            ))
        
        # Parse egress rules
        for rule in spec.get("egress", []):
            ports = parse_port_list(rule.get("ports", []))
            to_ns = parse_namespace_list(rule.get("to", []))
            
            policy_rules.append(PolicyRule(
                name=f"{name}-egress-{rule.get('name', 'unnamed')}",
                kind="Egress",
                ports=ports,
                from_namespaces=[],
                to_namespaces=to_ns
            ))
        
        return NetworkPolicy(
            name=name,
            namespace=namespace,
            kind="NetworkPolicy",
            spec=spec,
            rules=policy_rules
        )
    except Exception as e:
        print(f"Error parsing YAML: {e}")
        return None


def load_policies_from_file(filepath: str) -> List[NetworkPolicy]:
    """Load all NetworkPolicies from a file path."""
    filepath = Path(filepath).resolve()
    
    if not filepath.exists():
        raise FileNotFoundError(f"File not found: {filepath}")
    
    policies = []
    
    try:
        with open(filepath, "r") as f:
            content = f.read()
            
        # Try to parse as single policy first
        policy = load_policy_from_yaml(content)
        if policy:
            return [policy]
        
        # Otherwise assume multi-document YAML (list of policies)
        import yaml
        
        documents = list(yaml.safe_load_all(content))
        
        for doc in documents:
            if not isinstance(doc, dict):
                continue
                
            kind = doc.get("kind", "")
            if kind == "NetworkPolicy":
                policy = load_policy_from_yaml(yaml.dump([doc]))
                policies.append(policy)
                
    except yaml.YAMLError as e:
        raise ValueError(f"Invalid YAML format: {e}")
    
    return policies


# =============================================================================
# PORT & NAMESPACE PARSING HELPERS
# =============================================================================

def parse_port_list(ports_data: List[Dict]) -> List[PortRule]:
    """Parse Kubernetes port specification into PortRule objects."""
    ports = []
    
    for port_spec in ports_data:
        if not isinstance(port_spec, dict):
            continue
            
        protocol = port_spec.get("protocol", "TCP")
        
        # Handle named port (e.g., 80:http)
        port_str = str(port_spec.get("port"))
        if ":" in port_str:
            parts = port_str.rsplit(":", 1)
            try:
                port = int(parts[0])
                name = parts[1]
            except ValueError:
                continue
        else:
            try:
                port = int(port_spec.get("port", 80))
                name = str(port)
            except (ValueError, TypeError):
                continue
        
        ports.append(PortRule(
            name=name,
            protocol=protocol.upper(),
            port=port,
            action="Allow"
        ))
    
    return ports


def parse_namespace_list(nss_data: List[Dict]) -> List[NamespaceRule]:
    """Parse Kubernetes namespace specification into NamespaceRule objects."""
    namespaces = []
    
    for ns_spec in nss_data:
        if not isinstance(ns_spec, dict):
            continue
            
        # Handle named port (e.g., 80:http)
        ns_str = str(ns_spec.get("namespace"))
        
        # Parse namespace and optional selector
        parts = ns_str.rsplit(":", 1)
        try:
            name = int(parts[0])
            selector = parts[1] if len(parts) > 1 else None
        except ValueError:
            name = ns_spec.get("namespace", "default")
            selector = None
        
        namespaces.append(NamespaceRule(
            name=name,
            action="Allow"
        ))
    
    return namespaces


# =============================================================================
# VALIDATION LOGIC
# =============================================================================

class ValidationError(Exception):
    """Raised when a policy fails validation."""
    def __init__(self, message: str, severity: PolicyStatus = PolicyStatus.ERROR):
        super().__init__(message)
        self.severity = severity


def validate_policy(policy: NetworkPolicy) -> Tuple[bool, List[str]]:
    """Validate a single NetworkPolicy against best practices."""
    issues = []
    
    # Check for default deny (ingress or egress with no ports/namespaces)
    has_default_deny = policy.has_default_deny
    
    # Check rule complexity
    total_rules = sum(1 for r in policy.rules if r.kind == "Ingress") + \
                   sum(1 for r in policy.rules if r.kind == "Egress")
    
    if total_rules > MAX_RULES_PER_NAMESPACE:
        issues.append(f"Too many rules ({total_rules} > {MAX_RULES_PER_NAMESPACE})")
    
    # Check for overly permissive ingress (allows all namespaces)
    for rule in policy.rules:
        if rule.kind == "Ingress":
            if not rule.from_namespaces or any(
                ns.name == "*" or ns.name == "" 
                for ns in rule.from_namespaces
            ):
                issues.append(f"Rule {rule.name} allows ingress from all namespaces")
    
    # Check for overly permissive egress (allows all namespaces)
    for rule in policy.rules:
        if rule.kind == "Egress":
            if not rule.to_namespaces or any(
                ns.name == "*" or ns.name == "" 
                for ns in rule.to_namespaces
            ):
                issues.append(f"Rule {rule.name} allows egress to all namespaces")
    
    # Check for default ports that should be restricted
    default_ports_found = []
    for rule in policy.rules:
        if rule.kind == "Ingress":
            for port in rule.ports:
                if port.is_default_port and not port.name.startswith("http"):
                    default_ports_found.append(port)
    
    return len(issues) == 0, issues


def validate_namespace_policy(policies: List[NetworkPolicy]) -> AnalysisResult:
    """Validate a collection of policies against the security baseline."""
    result = AnalysisResult()
    
    for policy in policies:
        result.policies_analyzed += 1
        result.total_rules += sum(
            1 for r in policy.rules if r.kind == "Ingress"
        ) + sum(
            1 for r in policy.rules if r.kind == "Egress"
        )
        
        is_valid, issues = validate_policy(policy)
        
        if is_valid:
            result.compliant_policies += 1
            
        # Add findings
        for issue in issues:
            finding = {
                "policy": policy.name,
                "namespace": policy.namespace,
                "issue": issue,
                "severity": PolicyStatus.WARNING.value,
                "recommendation": get_recommendation(issue)
            }
            
            if "all namespaces" in issue.lower():
                finding["severity"] = PolicyStatus.ERROR.value
            
            result.findings.append(finding)
            result.warning_count += 1
    
    # Calculate final score
    result.score = result.compliant_policies / max(result.policies_analyzed, 1) * 100
    
    return result


def get_recommendation(issue: str) -> Optional[str]:
    """Get a remediation recommendation for an issue."""
    recommendations = {
        "all namespaces": f"Restrict to specific namespaces. Example:\n  from:\n  - namespaceSelector:\n      matchLabels:\n        name: <specific-namespace>",
        "default deny": "Add explicit default-deny rules at the end of your policy list.",
        "too many rules": "Consolidate similar rules or use label selectors for grouping.",
    }
    
    return recommendations.get(issue.lower().split()[0], None)


# =============================================================================
# GAP ANALYSIS & GENERATION
# =============================================================================

def analyze_gaps(policies: List[NetworkPolicy]) -> Dict[str, Any]:
    """Analyze gaps between current state and recommended baseline."""
    result = {
        "current_state": {
            "policies_count": len(policies),
            "total_rules": sum(
                1 for r in policy.rules if r.kind == "Ingress"
                for p in [policy] + policies
            )
        },
        "gaps": [],
        "recommended_policies": []
    }
    
    # Check for missing default-deny patterns
    has_default_deny_any = any(
        policy.has_default_deny for policy in policies
    )
    
    if not has_default_deny_any:
        result["gaps"].append({
            "type": "missing_default_deny",
            "description": "No explicit default-deny rules found across all policies",
            "severity": "HIGH"
        })
    
    # Check for namespace isolation gaps
    namespaces_used = set()
    for policy in policies:
        namespaces_used.add(policy.namespace)
        for rule in policy.rules:
            if rule.kind == "Ingress":
                for ns in rule.from_namespaces:
                    namespaces_used.add(ns.name)
            if rule.kind == "Egress":
                for ns in rule.to_namespaces:
                    namespaces_used.add(ns.name)
    
    # Check against recommended namespace isolation
    isolated_namespaces = [ns for ns in namespaces_used 
                          if not any("kube" in n.lower() or "default" in n.lower() 
                                    for n in DEFAULT_DENY_ALL_NAMESPACES)]
    
    result["gaps"].append({
        "type": "namespace_coverage",
        "description": f"Found {len(isolated_namespaces)} isolated namespaces: {', '.join(sorted(isolated_namespaces))}",
        "severity": "MEDIUM" if len(isolated_namespaces) < 3 else "LOW"
    })
    
    # Generate recommended policies
    result["recommended_policies"] = generate_recommended_policies(policies, namespaces_used)
    
    return result


def generate_recommended_policies(
    existing: List[NetworkPolicy], 
    namespaces: Set[str]
) -> List[Dict[str, Any]]:
    """Generate recommended NetworkPolicy YAML configurations."""
    recommendations = []
    
    # 1. Default-deny ingress policy (applied to each namespace)
    default_deny_ingress = {
        "apiVersion": "networking.k8s.io/v1",
        "kind": "NetworkPolicy",
        "metadata": {
            "name": "default-deny-ingress",
            "namespace": "<NAMESPACE>"  # User must replace this
        },
        "spec": {
            "podSelector": {},  # Apply to all pods
            "policyTypes": ["Ingress"],
            "ingress": []  # No ingress = deny all by default
        }
    }
    
    # 2. Default-deny egress policy (applied to each namespace)
    default_deny_egress = {
        "apiVersion": "networking.k8s.io/v1",
        "kind": "NetworkPolicy",
        "metadata": {
            "name": "default-deny-egress",
            "namespace": "<NAMESPACE>"  # User must replace this
        },
        "spec": {
            "pod